File: packages/websearch/src/engines/marginalia.ts
import type {
  WebSearchEngine,
  WebSearchEngineInput,
  WebSearchEngineResult,
  WebSearchResultItem,
} from "../types.js";
import { stripTags } from "./html.js";
import { httpGet } from "./http.js";
import { SearchError } from "./searchError.js";

const DEFAULT_BASE = "https://api.marginalia.nu";
const ENGINE_NAME = "marginalia";

/**
 * Marginalia public search API — keyless JSON, the cleanest ToS-wise of the
 * bundled keyless engines (a documented public API, not a scraped SERP).
 *
 *   GET https://api.marginalia.nu/public/search/{query}?count={n}
 *   → { license, query, results: [{ url, title, description, quality, ... }] }
 *
 * Maps title←title, url←url, snippet←description. The index is "small web"
 * (blogs, forums, docs) — excellent for technical/indie queries, weak on
 * mainstream and very-recent results. Results are CC-BY-NC-SA 4.0.
 *
 * @param opts.baseUrl override the API host (tests point this at a fixture
 *   server; production uses the default public host).
 */
export function createMarginaliaEngine(
  opts: { baseUrl?: string } = {},
): WebSearchEngine & { readonly name: string } {
  const base = opts.baseUrl ?? DEFAULT_BASE;
  return {
    name: ENGINE_NAME,
    async search(
      input: WebSearchEngineInput,
    ): Promise<WebSearchEngineResult> {
      const url = new URL(base);
      url.pathname = joinPath(url.pathname, [
        "public",
        "search",
        encodeURIComponent(input.query),
      ]);
      // Marginalia caps at ~100; ask for the model's count directly.
      url.searchParams.set("count", String(input.count));

      const res = await httpGet(url, input, {
        accept: "application/json",
        engine: ENGINE_NAME,
      });

      let parsed: unknown;
      try {
        parsed = JSON.parse(res.text);
      } catch (e) {
        throw new SearchError(
          "IO_ERROR",
          `marginalia: could not parse response as JSON: ${(e as Error).message}`,
          { engine: ENGINE_NAME },
        );
      }

      const results = mapResults(parsed);
      return { results, backendHost: res.host, elapsedMs: res.elapsedMs };
    },
  };
}

function mapResults(parsed: unknown): WebSearchResultItem[] {
  if (parsed === null || typeof parsed !== "object") return [];
  const raw = (parsed as { results?: unknown }).results;
  if (!Array.isArray(raw)) return [];
  const out: WebSearchResultItem[] = [];
  for (const entry of raw) {
    if (entry === null || typeof entry !== "object") continue;
    const e = entry as {
      title?: unknown;
      url?: unknown;
      description?: unknown;
    };
    const title = typeof e.title === "string" ? e.title : "";
    const url = typeof e.url === "string" ? e.url : "";
    if (title.length === 0 || url.length === 0) continue;
    const snippet =
      typeof e.description === "string" ? stripTags(e.description) : "";
    out.push({ title, url, snippet });
  }
  return out;
}

function joinPath(basePath: string, segments: string[]): string {
  const trimmed = basePath.replace(/\/+$/, "");
  return `${trimmed}/${segments.join("/")}`;
}

File: crates/websearch/src/engines/marginalia.rs
//! Marginalia public search API — keyless JSON. See TS `engines/marginalia.ts`.

use async_trait::async_trait;
use url::Url;

use super::html::strip_tags;
use super::http::http_get;
use crate::engine::{
    shared_client, SearchError, SearchErrorCode, WebSearchEngine, WebSearchEngineInput,
    WebSearchEngineResult,
};
use crate::types::{WebSearchResultItem, WebSearchTimeRange};

const DEFAULT_BASE: &str = "https://api.marginalia.nu";
const ENGINE_NAME: &str = "marginalia";

pub struct MarginaliaEngine {
    client: reqwest::Client,
    base_url: String,
}

impl MarginaliaEngine {
    pub fn new() -> Self {
        Self {
            client: shared_client(),
            base_url: DEFAULT_BASE.to_string(),
        }
    }
    pub fn with_base_url(mut self, base: impl Into<String>) -> Self {
        self.base_url = base.into();
        self
    }
}

impl Default for MarginaliaEngine {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl WebSearchEngine for MarginaliaEngine {
    fn name(&self) -> &str {
        ENGINE_NAME
    }

    fn engine_class(&self) -> crate::engine::EngineClass {
        crate::engine::EngineClass::Niche
    }

    async fn search(
        &self,
        input: WebSearchEngineInput,
    ) -> Result<WebSearchEngineResult, SearchError> {
        let mut url = Url::parse(&self.base_url).map_err(|_| {
            SearchError::new(
                SearchErrorCode::IoError,
                format!("invalid marginalia base url: {}", self.base_url),
            )
        })?;
        {
            let base_path = url.path().trim_end_matches('/').to_string();
            url.set_path(&format!("{}/public/search/{}", base_path, input.query));
            url.query_pairs_mut()
                .append_pair("count", &input.count.to_string());
        }

        let res = http_get(
            &self.client,
            &url,
            &input,
            "application/json",
            ENGINE_NAME,
            &[],
        )
        .await?;

        let parsed: serde_json::Value = serde_json::from_str(&res.text).map_err(|e| {
            SearchError::new(
                SearchErrorCode::IoError,
                format!("marginalia: could not parse response as JSON: {}", e),
            )
        })?;

        Ok(WebSearchEngineResult {
            results: map_results(&parsed),
            backend_host: res.host,
            elapsed_ms: res.elapsed_ms,
            engine: Some(ENGINE_NAME.to_string()),
            engine_class: None,
            engines: None,
            // Marginalia's public API has no recency filter.
            time_range_applied: if input.time_range == WebSearchTimeRange::All {
                None
            } else {
                Some(false)
            },
        })
    }
}

fn map_results(parsed: &serde_json::Value) -> Vec<WebSearchResultItem> {
    let raw = match parsed.get("results").and_then(|v| v.as_array()) {
        Some(arr) => arr,
        None => return Vec::new(),
    };
    let mut out = Vec::new();
    for entry in raw {
        let title = entry.get("title").and_then(|v| v.as_str()).unwrap_or("");
        let url = entry.get("url").and_then(|v| v.as_str()).unwrap_or("");
        if title.is_empty() || url.is_empty() {
            continue;
        }
        let snippet = entry
            .get("description")
            .and_then(|v| v.as_str())
            .map(strip_tags)
            .unwrap_or_default();
        // `quality` is Marginalia's relevance/quality score; surface it.
        let score = entry.get("quality").and_then(|v| v.as_f64());
        out.push(WebSearchResultItem {
            title: title.to_string(),
            url: url.to_string(),
            snippet,
            age: None,
            score,
            source: None,
        });
    }
    out
}
